### PR TITLE
Revamped preview diff highlight logic

### LINKED ISSUE #181: Preview jumps to the top of the page when editing inlined code
When editing inlined code in preview mode, instead of jumping to the change, webpage jumps to the top of the page.

### PR DESCRIPTION
**Pull request by MykolaGolubyev**
Overall approach:
* walk through dom nodes tree
* collect smallest textual content and associated dom nodes for rendered `before` html page and rendered `after` html page.
* diff two lists `before` and `after` using Myers algorithm (npm package). diff using only the textual content, and not any auxiliary information associated
* algo provides `added` and `deleted` elements
* go through `added` items and add an animation class to them

This PR also fixes image being constantly shown as updated.

Also closes #181

### CODE REVIEW

File: znai-reactjs/src/doc-elements/structure/tocItem.js [hunk 1]
```diff
@@ -14,14 +14,7 @@
...
-    animation-name: preview-change-keyframes;
-    animation-duration: 2s;
-    animation-fill-mode: both;
-}
-
-@keyframes preview-change-keyframes {
-    0% {background-color: rgba(38, 110, 164, 0)}
-    50% {background-color: rgba(38, 110, 164, 0.5)}
-    100% {background-color: rgba(38, 110, 164, 0)}
+export function areTocItemEquals(a, b) {
```
Comment (MykolaGolubyev) [inline znai-reactjs/src/doc-elements/structure/tocItem.js]: this diff looks weird, probably as I copied a file using IDEA and renamed it. But this is a new file.

File: znai-core/src/main/java/com/twosigma/znai/time/SystemTimeService.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    public long fileModifiedTimeMillis(Path path) {
+        try {
+            return Files.getLastModifiedTime(path).toMillis();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public long currentTimeMillis() {
```
Comment (byc1234) [inline znai-core/src/main/java/com/twosigma/znai/time/SystemTimeService.java]: Why do you need this method?

Comment (MykolaGolubyev) [reply znai-core/src/main/java/com/twosigma/znai/time/SystemTimeService.java]: As a future placeholder.

File: znai-core/src/test/groovy/com/twosigma/znai/time/FakeTimeService.groovy [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.twosigma.znai.time
+
+import java.nio.file.Path
+
+class FakeTimeService implements TimeService {
```
[the remark this answers is not in the document]
Comment (MykolaGolubyev) [reply znai-core/src/test/groovy/com/twosigma/znai/time/FakeTimeService.groovy]: I prefer using Fakes as they can be re-used across multiple tests.
In the same manner all the major components of znai were faked. 
Take a look at `ComponentsRegistry` and `TestComponentsRegistry`

File: znai-reactjs/src/diff/HtmlNodeDiff.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+        const isAnyChangeVisible = this.addedNodes.some(node => isNodeInViewport(nodeWithScroll, node))
+        if (isAnyChangeVisible) {
+            return
+        }
+
+        scrollIntoView(nodeWithScroll, this.addedNodes[0])
+    }
+
+    _extractAddedNodes() {
```
[the remark this answers is not in the document]
Comment (MykolaGolubyev) [reply znai-reactjs/src/diff/HtmlNodeDiff.js]: not at the moment

File: znai-core/src/test/groovy/com/twosigma/znai/time/FakeTimeService.groovy (hunk 3 above)
[the remark this answers is not in the document]
Comment (byc1234) [reply znai-core/src/test/groovy/com/twosigma/znai/time/FakeTimeService.groovy]: I also think that `TestComponentsRegistry` should be removed with a dependency injection. But we can refactor it later.

[the remark this answers is not in the document]
Comment (MykolaGolubyev) [reply znai-core/src/test/groovy/com/twosigma/znai/time/FakeTimeService.groovy]: I on purpose didn't bring any Dependency Injection framework. 
I get all the benefits of it with `ComponentsRegistry` without heavy weight of `Spring`. 
Let's not jump to refactoring yet :) We can chat more about it so you get a bigger picture.

[the remark this answers is not in the document]
Comment (byc1234) [reply znai-core/src/test/groovy/com/twosigma/znai/time/FakeTimeService.groovy]: You don't have to use DI framework like `Spring` or `Guice`, though they will help to reduce amount of code.
When you create an object you can just inject mock ones for testing like `ComponentsRegistry testComponentsRegistry = new ComponentsRegistry(mockDocStructure, mockMarkdownParser, ...);`.
Anyway we can talk about this later in person.

**Pull request merged**